plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE TupleSections      #-}
+{-# LANGUAGE ViewPatterns       #-}
+
+module PlutusTx.DataMap where
+
+import PlutusTx.Builtins qualified as P
+import PlutusTx.Builtins.Internal qualified as BI
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (null, toList)
+import PlutusTx.These
+
+import Prelude qualified as Haskell
+
+{- | An associative map implementation backed by `P.BuiltinData`.
+
+This map implementation has the following characteristics:
+
+  * The `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations are no-op.
+  * Other operations are slower than @PlutusTx.AssocMap.Map@, although equality
+    checks on keys can be faster due to `P.equalsData`.
+  * Many operations involve converting the keys and/or values to/from `P.BuiltinData`.
+
+Therefore this map implementation is likely a better choice than @PlutusTx.AssocMap.Map@
```

Comment:
Oh, are you going to keep both of them and let the user choose?  If you're going to do that you could maybe have property tests to check that both versions do the same thing.  Also in that case it would be worth mentioning it in `PlutusTx.AssocMap.Map`.